### ISSUE TITLE
[WPTableViewHandler.controllerDidChangeContent] NSInternalInconsistencyException: Attempted to perform update with invalid index path: <NSIndexPath: 0x2817c5fa0> {length = 2, path...

### ISSUE THREAD

**Issue by jkmassel** [wordpress-mobile/WordPress-iOS#15370]
**Sentry Url**: https://sentry.io/share/issue/26030d0c26be4d41b1bad2ce380e8b41/
**User Count**: 563
**Count**: 793
**First Release**: 15.5.0.3
**First Seen**: 2020-09-03T20:31:11Z
**Last Seen**: 2020-11-23T17:40:03Z
**24 Hours**: 167
**30 Days**: 714
**LogID:** 8A6ACD7C-E03A-4288-BAA9-E059D0DEF429

```
NSInternalInconsistencyException: Attempted to perform update with invalid index path: <NSIndexPath: 0x28295ab20> {length = 2, path...
  File "Classes/Utility/WPTableViewHandler.m", line 638, in -[WPTableViewHandler controllerDidChangeContent:]
  File "main.swift", line 7, in main
...
(34 additional frame(s) were not displayed)
```

**Comment by leandroalonso** [comment]
Duplicate of https://github.com/wordpress-mobile/WordPress-iOS/issues/15366

**Comment by yaelirub** [comment]
@leandroalonso , I don't think this is a duplicate of #15361 I actually just got this crash in Xcode, running on a simulator.

``` Terminating app due to uncaught exception 'NSInternalInconsistencyException', reason: '	: <NSIndexPath: 0x7fd981e366a0> {length = 2, path = 1 - 18446744073709551609}'
*** First throw call stack:
(
	0   CoreFoundation                      0x00007fff2041daf2 __exceptionPreprocess + 242
	1   libobjc.A.dylib                     0x00007fff20177e78 objc_exception_throw + 48
	2   CoreFoundation                      0x00007fff2041d91b +[NSException raise:format:] + 0
	3   Foundation                          0x00007fff2076b633 -[NSAssertionHandler handleFailureInFunction:file:lineNumber:description:] + 166
	4   UIKitCore                           0x00007fff248e746f _UIAssertValidUpdateIndexPath + 221
	5   UIKitCore                           0x00007fff248e733a -[UIUpdateItem initWithAction:forIndexPath:animation:] + 49
	6   UIKitCore                           0x00007fff2488d9fe -[UITableView _updateRowsAtIndexPaths:withUpdateAction:rowAnimation:usingPresentationValues:] + 550
	7   UIKitCore                           0x00007fff2488dbc5 -[UITableView insertRowsAtIndexPaths:withRowAnimation:] + 162
	8   WordPress                           0x00000001099387aa -[WPTableViewHandler controller:didChangeObject:atIndexPath:forChangeType:newIndexPath:] + 1482
	9   CoreData                            0x00007fff252c117b __82-[NSFetchedResultsController(PrivateMethods) _core_managedObjectContextDidChange:]_block_invoke + 6983
	10  CoreData                            0x00007fff2512b2ab developerSubmittedBlockToNSManagedObjectContextPerform + 154
	11  CoreData                            0x00007fff2512b192 -[NSManagedObjectContext performBlockAndWait:] + 197
	12  CoreData                            0x00007fff252bf622 -[NSFetchedResultsController _core_managedObjectContextDidChange:] + 110
	13  CoreFoundation                      0x00007fff2035ef02 __CFNOTIFICATIONCENTER_IS_CALLING_OUT_TO_AN_OBSERVER__ + 12
	14  CoreFoundation                      0x00007fff2035eed1 ___CFXRegistrationPost_block_invoke + 49
	15  CoreFoundation                      0x00007fff2035e5ef _CFXRegistrationPost + 454
	16  CoreFoundation                      0x00007fff2035df5d _CFXNotificationPost + 796
	17  Foundation                          0x00007fff207df98a -[NSNotificationCenter postNotificationName:object:userInfo:] + 59
	18  CoreData                            0x00007fff251b8029 -[NSManagedObjectContext _postObjectsDidChangeNotificationWithUserInfo:] + 474
	19  CoreData                            0x00007fff251c727c -[NSManagedObjectContext _createAndPostChangeNotification:deletions:updates:refreshes:deferrals:wasMerge:] + 1825
	20  CoreData                            0x00007fff251b901e -[NSManagedObjectContext _processRecentChanges:] + 1041
	21  CoreData                            0x00007fff251cc3b4 -[NSManagedObjectContext _parentProcessSaveRequest:inContext:error:] + 1563
	22  CoreData                            0x00007fff251cd0ff __82-[NSManagedObjectContext(_NestedContextSupport) executeRequest:withContext:error:]_block_invoke + 526
	23  CoreData                            0x00007fff2512b760 internalBlockToNSManagedObjectContextPerform + 89
	24  libdispatch.dylib                   0x0000000111d279c8 _dispatch_client_callout + 8
	25  libdispatch.dylib                   0x0000000111d36d63 _dispatch_async_and_wait_invoke + 175
	26  libdispatch.dylib                   0x0000000111d279c8 _dispatch_client_callout + 8
	27  libdispatch.dylib                   0x0000000111d35e75 _dispatch_main_queue_callback_4CF + 1152
	28  CoreFoundation                      0x00007fff2038bdab __CFRUNLOOP_IS_SERVICING_THE_MAIN_DISPATCH_QUEUE__ + 9
	29  CoreFoundation                      0x00007fff2038662e __CFRunLoopRun + 2685
	30  CoreFoundation                      0x00007fff203856c6 CFRunLoopRunSpecific + 567
	31  GraphicsServices                    0x00007fff2b76adb3 GSEventRunModal + 139
	32  UIKitCore                           0x00007fff24675187 -[UIApplication _run] + 912
	33  UIKitCore                           0x00007fff2467a038 UIApplicationMain + 101
	34  WordPress                           0x000000010a7bef64 main + 500
	35  libdyld.dylib                       0x00007fff20256409 start + 1
)
```

Whats the reason behind all those "Attempted to perform update with invalid index path" errors we're seeing? I tried to follow the discussion [here](https://github.com/wordpress-mobile/WordPress-iOS/issues/15368) but I'm not sure if its the same issue or not and what happened with that conversation

**Comment by leandroalonso** [comment]
@yaelirub do you have the steps to reproduce the crash?

I think they're not exactly happening at the same place, but all those crashes seem related to `WPTableViewHandler`. It started in November — and if you look into the OS versions being affected, they're almost 100% iOS 14.

So it might be that Core Data had some internal changes that are affecting us.

**Comment by designsimply** [comment]
In case it's relevant, I noticed a big spike in `NSInternalInconsistencyException` issues overall and they are currently showing up in [10 separate open issues in the repo](https://github.com/wordpress-mobile/WordPress-iOS/issues?q=is%3Aissue+is%3Aopen+NSInternalInconsistencyException)—I am not sure how closely those are related. I think the main contributor to the spike is #15366 though and the reason the issues might be getting split up like they are could be because of a recent change in how we group these types of issues in Sentry (explained in internal reference paaHJt-1jm-p2).

<a href="https://sentry.io/organizations/a8c/discover/results/?display=daily&field=timestamp&field=release&field=os&field=device&field=culprit&name=All+Events&project=1438083&query=NSInternalInconsistencyException&sort=-timestamp&statsPeriod=90d&widths=-1&widths=-1&widths=-1&widths=-1&widths=-1">![image](https://user-images.githubusercontent.com/1119271/101803344-a3751c80-3acd-11eb-9076-84860980b5dc.png)</a>

**Comment by AliSoftware** [comment]
Just for this crash today as well, just minutes ago.

- 23:56:58 — Received push notif (phone was locked)
- 00:01:52 — I tapped the push notification from my lock screen 
- App crashed immediately (I didn't even see it open in foreground, directly got the "app crashed, send logs?" alert
- 00:02:39 — I opened the app again (to access that logs)

(Full logs attached in #wp-crash-triage Slack thread in case it helps)

@guarani (as current groundskeeper): I think we should reopen the issue (or if there has been a different one opened for the same issue since, update the link in Sentry)

**Comment by guarani** [comment]
> Just for this crash today as well, just minutes ago.

I'm re-opening this issue because groundskeeping crash triage, which is timeboxed to 5-10 minutes, didn't give me enough time to verify whether this crash corresponds to this issue, https://github.com/wordpress-mobile/WordPress-iOS/issues/15366 (the issue this was marked a duplicate of), one of [the other `NSInternalInconsistencyException` issues](https://github.com/wordpress-mobile/WordPress-iOS/issues?q=is%3Aissue+is%3Aopen+NSInternalInconsistencyException), etc.

What I can say is that the Sentry issue your crash corresponds to, https://sentry.io/share/issue/26030d0c26be4d41b1bad2ce380e8b41, matches this issue, so it's likely that re-opening this is the right move.

**Comment by guarani** [comment]
I'm giving this Medium priority because:
- I couldn't reproduce using the [above steps](https://github.com/wordpress-mobile/WordPress-iOS/issues/15370#issuecomment-1034318724)
- It has a High severity due to being a crash
- It has a Low impact due to just one user report (from @AliSoftware) and no reliable way to reproduce the crash
- Using our priority matrix, this gives an overall Medium 

👋 Hi @ScoutHarris, I assigned this issue to you as part of groundskeeping crash triage. Please let me know if I should assign someone else to it.

**Comment by ScoutHarris** [comment]
Hey @guarani . @dvdchr has offered to take a look, so I reassigned it to him. Thanks!

**Comment by dvdchr** [comment]
👋  Spent a bit of time investigating this issue. Based on [Sentry](https://sentry.io/share/issue/26030d0c26be4d41b1bad2ce380e8b41/), there are currently two main sources for this crash:

### NotificationsViewController.selectRow

I've created a fix (attempt) here: #17955.  Note that the fix only added checks that ensure that we're working with a valid index path so the app is not crashing. This doesn't address the root cause yet, as I still have no idea when or why `NSFetchedResultsController` would return an index path with negative values.

### - [WPTableViewHandler controllerDidChangeContent]

Now, this is a trickier one, because the component is used in several places. Based on the stack trace, this happens after a call to `[tableView endUpdates]` which leads to an exception due to invalid index paths. Based on the Sentry breadcrumbs, most of the crash instances show that the crash happened after the app entered the background state, although I did manage to find a crash instance that happens in the foreground state.

I did notice #15475 where it solves a very similar issue to this one (also in `WPTableViewHandler`) – and I think it works. What's more, #15475 also includes `controllerDidChangeContent`. However, the crash happens in `endUpdates`, just right before the fix:

https://github.com/wordpress-mobile/WordPress-iOS/blob/bd8121235d0638da420461397f61608c1be287e7/WordPress/Classes/Utility/WPTableViewHandler.m#L627-L637

However, I'm not sure if we can just simply skip `endUpdates`, because `beginUpdates` is called in `controllerWillChangeContent`. We'll need to take an even closer look at this one since it could impact multiple screens.

**Comment by salimbraksa** [comment]
> - [WPTableViewHandler controllerDidChangeContent]

Closing this issue as it is resolved in [PR](https://github.com/wordpress-mobile/WordPress-iOS/pull/21250).

> NotificationsViewController.selectRow

This is an ongoing issue and tracked in a separate [Github issue](https://github.com/wordpress-mobile/WordPress-iOS/issues/15368).
